File: dexfile/src/main/java/com/github/netomi/bat/dexfile/io/DataSectionItemCollector.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.github.netomi.bat.dexfile.io;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.annotation.*;
import com.github.netomi.bat.dexfile.debug.DebugInfo;
import com.github.netomi.bat.dexfile.visitor.DataItemVisitor;

/**
 * @author Thomas Neidhart
 */
public class DataSectionItemCollector
implements   DataItemVisitor
{
    private DexFileWriter.DataItemMap dataItemMap;

    public DataSectionItemCollector(DexFileWriter.DataItemMap dataItemMap) {
        this.dataItemMap = dataItemMap;
    }

    @Override
    public void visitAnyDataItem(DexFile dexFile, DataItem dataItem) {}

    @Override
    public void visitStringData(DexFile dexFile, StringID stringID, StringData stringData) {
        dataItemMap.addDataItem(stringData);
    }

    @Override
    public void visitParameterTypes(DexFile dexFile, ProtoID protoID, TypeList typeList) {
        dataItemMap.addDataItem(typeList);
    }

    @Override
    public void visitClassData(DexFile dexFile, ClassDef classDef, ClassData classData) {
        dataItemMap.addDataItem(classData);
    }

    @Override
    public void visitInterfaceTypes(DexFile dexFile, ClassDef classDef, TypeList typeList) {
        dataItemMap.addDataItem(typeList);
    }

    @Override
    public void visitStaticValuesArray(DexFile dexFile, ClassDef classDef, EncodedArray encodedArray) {
        dataItemMap.addDataItem(encodedArray);
    }

    @Override
    public void visitAnnotationsDirectory(DexFile dexFile, ClassDef classDef, AnnotationsDirectory annotationsDirectory) {
        dataItemMap.addDataItem(annotationsDirectory);
    }

    @Override
    public void visitClassAnnotations(DexFile dexFile, AnnotationsDirectory annotationsDirectory, AnnotationSet annotationSet) {
        dataItemMap.addDataItem(annotationSet);
    }

    @Override
    public void visitFieldAnnotations(DexFile dexFile, FieldAnnotation fieldAnnotation, AnnotationSet annotationSet) {
        dataItemMap.addDataItem(annotationSet);
    }

    @Override
    public void visitMethodAnnotations(DexFile dexFile, MethodAnnotation methodAnnotation, AnnotationSet annotationSet) {
        dataItemMap.addDataItem(annotationSet);
    }

    @Override
    public void visitParameterAnnotations(DexFile dexFile, ParameterAnnotation parameterAnnotation, AnnotationSetRefList annotationSetRefList) {
        dataItemMap.addDataItem(annotationSetRefList);
    }

    @Override
    public void visitAnnotationSet(DexFile dexFile, AnnotationSetRef annotationSetRef, AnnotationSet annotationSet) {
        dataItemMap.addDataItem(annotationSet);
    }

    @Override
    public void visitAnnotation(DexFile dexFile, AnnotationSet annotationSet, int index, Annotation annotation) {
        dataItemMap.addDataItem(annotation);
    }

    @Override
    public void visitCode(DexFile dexFile, EncodedMethod encodedMethod, Code code) {
        dataItemMap.addDataItem(code);
    }

    @Override
    public void visitDebugInfo(DexFile dexFile, Code code, DebugInfo debugInfo) {
        dataItemMap.addDataItem(debugInfo);
    }

    @Override
    public void visitCallSite(DexFile dexFile, CallSiteID callSiteID, CallSite callSite) {
        dataItemMap.addDataItem(callSite);
    }
}

File: dexfile/src/main/java/com/github/netomi/bat/dexfile/DataSectionItemCollector.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile

import com.github.netomi.bat.dexfile.annotation.*
import com.github.netomi.bat.dexfile.annotation.Annotation
import com.github.netomi.bat.dexfile.debug.DebugInfo
import com.github.netomi.bat.dexfile.visitor.DataItemVisitor

internal class DataSectionItemCollector(private val dataItemMap: DataItemMapImpl) : DataItemVisitor {

    override fun visitAnyDataItem(dexFile: DexFile, dataItem: DataItem) {}

    override fun visitStringData(dexFile: DexFile, stringID: StringID, stringData: StringData) {
        dataItemMap.addDataItem(stringData)
    }

    override fun visitParameterTypes(dexFile: DexFile, protoID: ProtoID, typeList: TypeList) {
        dataItemMap.addDataItem(typeList)
    }

    override fun visitClassData(dexFile: DexFile, classDef: ClassDef, classData: ClassData) {
        dataItemMap.addDataItem(classData)
    }

    override fun visitInterfaceTypes(dexFile: DexFile, classDef: ClassDef, typeList: TypeList) {
        dataItemMap.addDataItem(typeList)
    }

    override fun visitStaticValuesArray(dexFile: DexFile, classDef: ClassDef, encodedArray: EncodedArray) {
        dataItemMap.addDataItem(encodedArray)
    }

    override fun visitAnnotationsDirectory(dexFile: DexFile, classDef: ClassDef, annotationsDirectory: AnnotationsDirectory) {
        dataItemMap.addDataItem(annotationsDirectory)
    }

    override fun visitClassAnnotations(dexFile: DexFile, annotationsDirectory: AnnotationsDirectory, annotationSet: AnnotationSet) {
        dataItemMap.addDataItem(annotationSet)
    }

    override fun visitFieldAnnotations(dexFile: DexFile, fieldAnnotation: FieldAnnotation, annotationSet: AnnotationSet) {
        dataItemMap.addDataItem(annotationSet)
    }

    override fun visitMethodAnnotations(dexFile: DexFile, methodAnnotation: MethodAnnotation, annotationSet: AnnotationSet) {
        dataItemMap.addDataItem(annotationSet)
    }

    override fun visitParameterAnnotations(dexFile: DexFile, parameterAnnotation: ParameterAnnotation, annotationSetRefList: AnnotationSetRefList) {
        dataItemMap.addDataItem(annotationSetRefList)
    }

    override fun visitAnnotationSet(dexFile: DexFile, annotationSetRef: AnnotationSetRef, annotationSet: AnnotationSet) {
        dataItemMap.addDataItem(annotationSet)
    }

    override fun visitAnnotation(dexFile: DexFile, annotationSet: AnnotationSet, index: Int, annotation: Annotation) {
        dataItemMap.addDataItem(annotation)
    }

    override fun visitCode(dexFile: DexFile, encodedMethod: EncodedMethod, code: Code) {
        dataItemMap.addDataItem(code)
    }

    override fun visitDebugInfo(dexFile: DexFile, code: Code, debugInfo: DebugInfo) {
        dataItemMap.addDataItem(debugInfo)
    }

    override fun visitCallSite(dexFile: DexFile, callSiteID: CallSiteID, callSite: CallSite) {
        dataItemMap.addDataItem(callSite)
    }
}